vms/components/fee/validator.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fee
+
+import (
+	"fmt"
+
+	"github.com/holiman/uint256"
+)
+
+type ValidatorState struct {
+	Current  Gas
+	Target   Gas
+	Capacity Gas
+	Excess   Gas
+	MinFee   GasPrice
+	K        Gas
+}
+
+func (v ValidatorState) CurrentFeeRate() GasPrice {
+	return v.MinFee.MulExp(v.Excess, v.K)
+}
+
+func (v ValidatorState) CalculateContinuousFee(seconds uint64) uint64 {
+	if v.Current == v.Target {
+		return uint64(v.MinFee.MulExp(v.Excess, v.K)) * seconds
+	}
+
+	uint256.NewInt(uint64(v.Excess))
+
+	var totalFee uint64
+	if v.Current < v.Target {
+		secondsTillExcessIsZero := uint64(v.Excess / (v.Target - v.Current))
+
+		if secondsTillExcessIsZero < seconds {
+			totalFee += uint64(v.MinFee) * (seconds - secondsTillExcessIsZero)
+			seconds = secondsTillExcessIsZero
+		}
+	}
+
+	x := v.Excess
+	for i := uint64(1); i <= seconds; i++ {
```

Comment:
This is more standard (because `i` can never overflow)
```suggestion
	for i := uint64(0); i < seconds; i++ {
```